ql/src/java/org/apache/hadoop/hive/ql/metadata/Table.java

```diff
@@ -1126,71 +1149,167 @@ public Boolean isOutdatedForRewriting() {
     return outdatedForRewritingMaterializedView;
   }
 
-  /* These are only populated during optimization and describing */
+  public ColumnStatistics getColStats() {
+    return tTable.isSetColStats() ? tTable.getColStats() : null;
+  }
+
+  /**
+   * Setup the table level stats as if the table is new. Used when setting up Table for a new
+   * table or during replication.
+   */
+  public void setStatsStateLikeNewTable() {
+    if (isPartitioned()) {
+      StatsSetupConst.setStatsStateForCreateTable(getParameters(), null,
+          StatsSetupConst.FALSE);
+    } else {
+      StatsSetupConst.setStatsStateForCreateTable(getParameters(),
+          MetaStoreUtils.getColumnNames(getCols()), StatsSetupConst.TRUE);
+    }
+  }
+
+  /** Constraints related methods
+   *  Note that set apis are used by DESCRIBE only, although get apis return RELY or ENABLE
+   *  constraints DESCRIBE could set all type of constraints
+   * */
+
+  /* This only return PK with RELY */
   public PrimaryKeyInfo getPrimaryKeyInfo() {
+    if(!this.isPKFetched) {
+      try {
+        pki = Hive.get().getReliablePrimaryKeys(this.getDbName(), this.getTableName());
+        this.isPKFetched = true;
+      } catch (HiveException e) {
+        LOG.warn("Cannot retrieve PK info for table : " + this.getTableName()
+            + " ignoring exception: " + e);
+      }
+    }
     return pki;
   }
 
   public void setPrimaryKeyInfo(PrimaryKeyInfo pki) {
     this.pki = pki;
+    this.isPKFetched = true;
   }
 
+  /* This only return FK with RELY */
   public ForeignKeyInfo getForeignKeyInfo() {
```

Comment:
This solution will still make different calls to retrieve each type of constraints. Should we create a follow-up for that?

Reply:
Created https://issues.apache.org/jira/browse/HIVE-22782 for follow-up